### DISCUSSION TITLE
Medium/Large OSS projects with roda/sequel

### DISCUSSION THREAD

**Discussion by ineb** [jeremyevans/roda#274] [General]
Hi all,
my own Roda/Sequel application is growing in size and i have some problems in keeping my project structure organized. 

Do you know any well-structured medium/large-sized projects using the roda/sequel stack which one can use for inspiration about project/code structure?

Thanks for any recommendations :)

**Comment by adam12** [comment]
I have a few medium-largish applications. They're reasonably organized (IMHO), but I'm guessing there is still some room for improvement. You didn't mention specifically what issues you were having, but here are a few design decisions I've made as it's grown:

1. Vertical "slices" for reasonable contexts (ala Merb or Hanami Slices). I have a "slice" for the Admin section, a "slice" for the Main section, and a "slice" for the API. I have a Router "slice" that is the entry point from Rack that handles the routing to the slices.

2. More route files. It's difficult because naming things is one of my least favourite things, but generally, I find large route files to be one of the most common issues that other teammates have. They become lost in large routes but are afraid to break things up.

3. No more mystery guest instance variables. I'm still working on removing these in some of my larger applications, but anything shared across route files is shared with the `shared_vars` plugin. I want almost no instance variables for route state as it becomes a mystery who depends on what. At the top of the route file, I extract the shared values I need into locals, and it's clear when looking at the routes what collaborators are required and provided from elsewhere.

4. Forced subfolders for views. I actually started on a plugin to enable this, but I don't remember how far I completed it. Essentially, if you give developers the option to just stick files anywhere, they will. For views, this becomes a bit of a dumping ground. If it's a routes file with a name, it should almost always have a view subdir forced upon it.

5. Everything through `view` or `render` is a local, unless it's something known to be application-specific such as CSRF tag methods.

One thing I miss is the ability to have route-specific methods, without cluttering the entire application. I've done a few approaches, and I'm not really sure the direction I like yet. A few that I've tried are:

1. "Page" objects that contain methods and values for the view being rendered.
2. dry-view / hanami-view.
3. Passing everything as locals, and lambdas for lazy state.

I've seen some people suggesting subclassing from `Roda` or `App` for each route file, which is intriguing and something I might explore as well, tho I suspect it's not perfect.

Hope this helps.

**Comment by jeremyevans** [comment]
My largest application has about 50 route branches and 150 models, so still very much a medium-sized application and not a large application.  It's structured basically the same as roda-sequel-stack (https://github.com/jeremyevans/roda-sequel-stack), with one routing file per branch.

In comparison to @adam12:

* No slices.  Only a single Roda application for the whole app.

* No major issues with route file size.  About 7K lines in the route files, but the largest route file only has about 750 lines, which is still easily manageable.

* Instance variables used for almost all state (never had problems with this approach)

* View subdirectories for each route file, using the new `hash_branch_view_subdir` plugin (this will be in the next release).

* Instance variables used to pass data to `view`/`render`, locals only used in special cases.

I have few route-specific methods in my application, and all are due to code written before the switch to Roda (the application was converted from Rails to Roda in 2014).  A lot of routes use route-specific data (local variables declared before the hash branch).

Like @adam12 said, it would be easier to provide more specific advice if you let us know what specific issues you are having.

**Comment by ineb** [comment]
Thank you both for your great suggestions. Can you provide links to your mentioned repositories?

Sorry for not being more specific in my entry post. I thought it would be easier if i could search for solutions in existing projects instead of discussing details here but both of you already addressed my main problems perfectly.

I see route files getting larger and more complex in my project. "Mystery guest instance variables" are indeed another problem. The state of various variables is getting unclear in my project and i think the mentioned `shared_vars` is able to solve this issue for me.

Besides that, i always like to read code of external projects with the same technology stack for inspiration. I almost always find something that leads to improvement in my own code.

**Reply by adam12** [reply]
> Can you provide links to your mentioned repositories?

Unfortunately not. My projects of the medium+ variety are all proprietary, some for clients.

**Reply by jeremyevans** [reply]
I linked to roda-sequel-stack above.  Like @adam12, my larger applications are not open source.

**Reply by ineb** [reply]
No problem, thanks. roda-sequel-stack is an excellent template to start from.
